File: 53-maximum-subarray/53-maximum-subarray.cpp
class Solution {
public:
    int ans = 0;
    long long int memo[100000+5];
    int dp(int pos,vector<int>& ar)
    {
        if(pos==0)
        {
            ans=max(ans,ar[pos]);
            return ar[pos];
        }
        
        if(memo[pos]!=-1)
            return memo[pos];
        
        
        ans=max(ans,max(ar[pos]+dp(pos-1,ar),ar[pos]));
        
        return memo[pos]=max(ar[pos]+dp(pos-1,ar),ar[pos]);
    }
    int maxSubArray(vector<int>& nums) {
        ans=nums[0];
        memset(memo,-1,sizeof(memo));

        dp(nums.size()-1,nums);
        return ans;
    }
};

File: 53-maximum-subarray/53-maximum-subarray.py
class Solution:
    def maxSubArray(self, ar: List[int]) -> int:
        
        
        def mxcro(st,mid,ed):
            
            tot,lsum,rsum = 0,float(-inf),float(-inf)
            
            for i in range(mid,st-1,-1):
                tot+=ar[i]
                lsum = max(lsum,tot)
            
            tot = 0;
            
            for i in range(mid+1,ed+1):
                tot+=ar[i]
                rsum = max(rsum,tot)
    
            return max(lsum,max(rsum,lsum+rsum))
        
        def mxsub(st,ed):
            if st==ed:
                return ar[st]
            mid = (st+ed)//2
            return max(mxsub(st,mid),max(mxsub(mid+1,ed),mxcro(st,mid,ed)))
        
        return mxsub(0,len(ar)-1)
